### ISSUE TITLE
Kubelet does not sync pod updates for static pods

### ISSUE THREAD

**Issue by vinaykul** [kubernetes/kubernetes#116597]
### What happened?

When a static pod is created on a node, it's mirror pod is created on the apiserver. Patching the mirror pod on the apiserver triggers HandlePodUpdate function in the kubelet:
https://github.com/kubernetes/kubernetes/blob/fcf5d23e6818ccfbc92f332af82f31661f8d4f88/pkg/kubelet/kubelet.go#L2491-L2502
However, handleMirrorPod sets up SyncPodUpdate for the original pod instead of the updated pod.
https://github.com/kubernetes/kubernetes/blob/fcf5d23e6818ccfbc92f332af82f31661f8d4f88/pkg/kubelet/kubelet.go#L2412-L2420
As a result, SyncPod does not really process the updates for static pods.

### What did you expect to happen?

Pod updates should be processed for static pods in the same manner as they are processed for normal pods.

### How can we reproduce it (as minimally and precisely as possible)?

Start local cluster w/ latest master (as of Mar 14, 2023):
```bash
# echo > /tmp/kubelet.log && DBG=1 ./hack/local-up-cluster.sh
...
...
+++ [0314 14:50:37] WARNING: linux/arm will no longer be built/shipped by default, please build it explicitly if needed.
+++ [0314 14:50:37]          support for linux/arm will be removed in a subsequent release.
make: Entering directory '/root/go/src/k8s.io/master_k8s'
+++ [0314 14:50:37] WARNING: linux/arm will no longer be built/shipped by default, please build it explicitly if needed.
+++ [0314 14:50:37]          support for linux/arm will be removed in a subsequent release.
go version go1.20.2 linux/arm64
+++ [0314 14:50:38] Building go targets for linux/arm64
    k8s.io/kubernetes/cmd/kubectl (static)
    k8s.io/kubernetes/cmd/kube-apiserver (static)
...
...
To start using your cluster, you can open up another terminal/tab and run:

  export KUBECONFIG=/var/run/kubernetes/admin.kubeconfig
  cluster/kubectl.sh

Alternatively, you can write to the default kubeconfig:

  export KUBERNETES_PROVIDER=local

  cluster/kubectl.sh config set-cluster local --server=https://localhost:6443 --certificate-authority=/var/run/kubernetes/server-ca.crt
  cluster/kubectl.sh config set-credentials myself --client-key=/var/run/kubernetes/client-admin.key --client-certificate=/var/run/kubernetes/client-admin.crt
  cluster/kubectl.sh config set-context local --cluster=local --user=myself
  cluster/kubectl.sh config use-context local
  cluster/kubectl.sh
```

Create static pod:
```bash
# cat <<EOF > /var/run/kubernetes/static-pods/bbox.yaml
apiVersion: v1
kind: Pod
metadata:
  name: bbox
spec:
  restartPolicy: OnFailure
  terminationGracePeriodSeconds: 2
  containers:
  - name: ctr
    image: busybox:1.33
    imagePullPolicy: IfNotPresent
    command: ["tail", "-f", "/dev/null"]
EOF
```

Verify image in container status of the static pod:
```bash
# ./cluster/kubectl.sh get pod bbox-127.0.0.1 -ojson | jq .status.containerStatuses[0].image
"docker.io/library/busybox:1.33"
```

Patch container image:
```bash
# ./cluster/kubectl.sh patch pod bbox-127.0.0.1 --patch '{"spec":{"containers":[{"name":"ctr", "image":"busybox:1.34"}]}}'
pod/bbox-127.0.0.1 patched
```

Check pod spec of static pod and verify image has been updated:
```bash
# ./cluster/kubectl.sh get pod bbox-127.0.0.1 -ojson | jq .spec.containers[0].image
"busybox:1.34"
```

After some time, check container status of static pod:
```bash
# ./cluster/kubectl.sh get pod bbox-127.0.0.1 -ojson | jq .status.containerStatuses[0].image
"docker.io/library/busybox:1.33"
``` 

Repeat exercise with normal pod to verify that container restarts and image has been updated.

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
# ./cluster/kubectl.sh version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"27+", GitVersion:"v1.27.0-alpha.3.573+a9008b502d32a5", GitCommit:"a9008b502d32a5dbacf835adb040419bf31f5c04", GitTreeState:"clean", BuildDate:"2023-03-14T14:50:38Z", GoVersion:"go1.20.2", Compiler:"gc", Platform:"linux/arm64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"27+", GitVersion:"v1.27.0-alpha.3.573+a9008b502d32a5", GitCommit:"a9008b502d32a5dbacf835adb040419bf31f5c04", GitTreeState:"clean", BuildDate:"2023-03-14T14:50:38Z", GoVersion:"go1.20.2", Compiler:"gc", Platform:"linux/arm64"}
```
</details>

### Cloud provider

<details>
Local ubuntu VM on mac M1
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
# uname -a
Linux vbuild 5.4.0-144-generic #161-Ubuntu SMP Fri Feb 3 14:48:57 UTC 2023 aarch64 aarch64 aarch64 GNU/Linux
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by vinaykul** [comment]
IIUC, this is not even a bug, in light of info that static pods [are not expected to be resized via API](https://github.com/kubernetes/kubernetes/pull/116504#issuecomment-1468696150) , which I was unsure of when I created this issue. @smarterclayton please confirm.

That said, Imho it might be desirable to resize under/over provisioned static pods based on observed usage for better cluster utilization.

**Comment by smarterclayton** [comment]
Yes, the source of truth for spec of a static pod is the config on the node, not the mirror pod.  There is no component today that reconciles the state of the pod spec on mirror pods, which would be part of #16627 through David's issue.  The HandlePodUpdate is triggered to indicate that the status of the pod has changed, which is part of the flow of status info from kubelet -> api -> back to pod worker.

**Comment by Vandit1604** [comment]
Hey, I searched for the above code in the master branch and it has been removed by this PR https://github.com/kubernetes/kubernetes/pull/102884. 
Also [here](https://kubernetes.io/docs/tasks/configure-pod-container/static-pod/#dynamic-addition-and-removal-of-static-pods), it states that you can only update something in a static pod if the configured directory has some changes in it. Through the `api-server`. It's not possible to update via changing something in mirror-pod.

Also, I'm new to k8s and finding my first issue to work on, if my deduction in wrong in anyway. Please feel free to point it out (Happy to take brutal feedback)

**Comment by tallclair** [comment]
We need to be very careful about handling any updates from the apiserver for mirror pods, as there is potential for a security issues compromising node isolation here. Mirror pods are the only type of pods that nodes are allowed to create, and in the past we've had some close calls that were mitigated by the fact that Kubelets basically ignore mirror pods.

**Comment by ndixita** [comment]
@Vandit1604 Would you like to help us here?

**Comment by tallclair** [comment]
The mirror pod is just a reflection of the static pod, not a source of the truth. The source of truth for the static pod spec is the static pod source, and any changes to that cause the pod to be recreated. IMO it's not worth handling in-place updates to the static pod source, but if we wanted to implement this, I think that's where it would need to go.

For now, we disallow mirror pod resizes in validation, which I think is the correct behavior: https://github.com/kubernetes/kubernetes/blob/3a1157f29e148f8212f23ce306d214743e49d1d7/pkg/apis/core/validation/validation.go#L5652-L5655

/close
